python/singa/layer.py

```diff
@@ -8,1443 +8,1279 @@
-        _check_engine(engine, ['cudnn', 'singacpp', 'singacl'])
-        self.layer = _create_layer(engine, 'Pooling')
-        if input_sample_shape is not None:
-            self.setup(input_sample_shape)
+    def initialize(self, *input):
+        pass
 
-    def setup(self, in_shape):
-        '''Set up the kernel, stride and padding; then call the C++ setup
-        function to create params and set some meta data.
+    def forward(self, *input):
+        pass
 
-        Args:
-            in_shapes is a tuple of int for the input sample shape
-        '''
-        if self.has_setup:
-            return
-        _set_kernel_stride_pad(self.conf.pooling_conf, self.kernel, self.stride,
-                               self.border_mode, self.pad, in_shape)
-        self.layer.Setup(list(in_shape), self.conf.SerializeToString())
-        self.has_setup = True
+    def __call__(self, *args, **kwargs):
+        if not self._initialized:
+            self.initialize(*args, **kwargs)
+            self._initialized = True
+
+        return self.forward(*args, **kwargs)
+
+    def get_params(self):
+        params = dict()
+        sublayers = self._layers
+        for name, sublayer in sublayers.items():
+            params.update(sublayer.get_params())
+        return params
+
+    def set_params(self, parameters):
+        # set parameters for Layer
+        # input should be either a PyTensor or numpy ndarray.
+        # examples: Layer.set_params(W=np.ones((in, out), dtype=np.float32)),
+        # Layer.set_params(**{'block1':{'linear1':{'W':np.ones((in, out),
+        # dtype=np.float32)}}})
+        sublayers = self._layers
+        for name, sublayer in sublayers.items():
+            sublayer.set_params(parameters)
+
+    def get_states(self):
+        states = dict()
+        sublayers = self._layers
+        for name, sublayer in sublayers.items():
+            states.update(sublayer.get_states())
+        states.update(self.get_params())
+        return states
+
+    def set_states(self, states):
+        sublayers = self._layers
+        for name, sublayer in sublayers.items():
+            sublayer.set_states(states)
+        self.set_params(states)
+
+    def device_check(self, *inputs):
+        x_device = inputs[0].device
+        x_dev_id = x_device.id()
+        for var in inputs:
+            if var.device.id() != x_dev_id:
+                var.to_device(x_device)
+
+    def set_attribute(self, attribute, attribute_value):
+        assert (attribute_value.shape == attribute.shape), "Shape dismatched."
+        if isinstance(attribute_value, Tensor):
+            attribute.reset_like(attribute_value)
+            attribute.copy_data(attribute_value)
+        elif isinstance(attribute_value, np.ndarray):
+            attribute.copy_from_numpy(attribute_value)
+        else:
+            raise ValueError("attributes should be Tensor or Numpy array.")
+
+    def _get_unique_name(self):
+        prefix = ''
```

Comment:
this function is be called recursively to set all sublayers' name?

Reply:
Not yet. It just gets the name of the parent layer of the current layer, and then adds the separator to become the name of the current layer. Since we update name by calling the initialize function and all layers are initialized from top to bottom, so we can ensure that each name is unique. 
https://github.com/apache/singa/pull/697/files#diff-3240fd3b573c3eaf8f8d6da8470e68e7R46

Reply:
got it. I just found it is called by the wrapper func.